From 6a7ba89884375573a8ffc971d1460011648aebaa Mon Sep 17 00:00:00 2001
From: Suraj P Kizhakkethil <quic_surapk@quicinc.com>
Date: Mon, 30 Dec 2024 17:00:18 +0530
Subject: [PATCH] QSDK:CP:wifi-scripts: Fix mac address generation for 2GHz,
 5GHz and sta

Currently, after openwrt upgrade, the mac address generation method
for 2GHz, 5GHz and sta is incorrect. mbssid is set for sta and this results in incorrect address generation.
Add check for mode when setting mbssid. Set base_mask to ff:ff:ff:ff:ff:ff when base_mask is 00:00:00:00:00:00 to match previous openwrt and qsdk behaviour.

Change-Id: I71df72047b448c75932d0cef8bcebc6b567b66c5
Signed-off-by: Suraj P Kizhakkethil <quic_surapk@quicinc.com>
---
 .../files/lib/netifd/wireless/mac80211.sh           | 13 ++++++++++---
 .../wifi-scripts/files/usr/share/hostap/common.uc   |  3 +--
 2 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh b/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
index af2c88c2f0b9..515bf5437613 100755
--- a/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
+++ b/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
@@ -928,13 +928,19 @@ mac80211_generate_mac() {
 	local phy="$1"
 	local id="${macidx:-0}"
 	local group_size
+	local mode="$2"
+	local mbssid=""
 	if [ -n "mbssid_group_size" ]; then
 		group_size="$mbssid_group_size"
 	else
 		group_size="$has_ap"
 	fi
 
-	wdev_tool "$phy$phy_suffix" get_macaddr id=$id num_global=$num_global_macaddr mbssid=$multiple_bssid mbssid_group_size=$group_size macaddr_base=${macaddr_base}
+	if [ "$mode" = "ap" ]; then
+		mbssid=$multiple_bssid
+	fi
+
+	wdev_tool "$phy$phy_suffix" get_macaddr id=$id num_global=$num_global_macaddr mbssid=$mbssid mbssid_group_size=$group_size macaddr_base=${macaddr_base}
 }
 
 get_board_phy_name() (
@@ -1102,7 +1108,7 @@ mac80211_prepare_vif() {
 
 	default_macaddr=
 	if [ -z "$macaddr" ]; then
-		macaddr="$(mac80211_generate_mac $phy)"
+		macaddr="$(mac80211_generate_mac $phy $mode)"
 		macidx="$(($macidx + 1))"
 		#default_macaddr=1
 	elif [ "$macaddr" = 'random' ]; then
@@ -1809,7 +1815,8 @@ drv_mac80211_setup() {
 
 	mac80211_prepare_iw_htmode
 	active_ifnames=
-	for_each_interface "ap sta adhoc mesh monitor" mac80211_prepare_vif
+	for_each_interface "ap" mac80211_prepare_vif
+	for_each_interface "sta adhoc mesh monitor" mac80211_prepare_vif
 	for_each_interface "ap sta adhoc mesh monitor" mac80211_setup_vif
 
 	[ -x /usr/sbin/wpa_supplicant ] && wpa_supplicant_set_config "$phy" "$radio"
diff --git a/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc b/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc
index 81ba48bd4c11..cc568984afb0 100644
--- a/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc
+++ b/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc
@@ -232,6 +232,7 @@ const phy_proto = {
 				if (idx < length(addrs))
 					return addrs[idx];
 			}
+			base_mask = "ff:ff:ff:ff:ff:ff";
 		}
 
 		if (!idx && !mbssid)
-- 
2.34.1

